File: python/DataStructures/CircularQueue/CircularQueue.py
from typing import List, TypeVar, Generic

T = TypeVar('T')


class CircularQueue(Generic[T]):
    capacity: int
    data: List[T or None]
    front_pointer: int
    rear_pointer: int
    size: int

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.data = [None]*capacity
        self.front_pointer = 0
        self.rear_pointer = 0
        self.size = 0

    def __str__(self):
        return "{}".format(self.data)

    def is_full(self):
        return self.size == self.capacity

    def is_empty(self):
        return self.size == 0

    def enqueue(self, item: T):
        if self.is_full():
            raise Exception('Queue Full')

        self.data[self.rear_pointer] = item
        self.rear_pointer += 1
        self.rear_pointer %= self.capacity
        self.size += 1

    def dequeue(self) -> T:
        if self.is_empty():
            raise Exception('Queue Empty')

        item: T = self.data[self.front_pointer]
        self.front_pointer += 1
        self.front_pointer %= self.capacity
        self.size -= 1
        return item

File: lib/src/dataStructures/queue/CircularQueue.ts
import { ObservableVersioned } from "common";

export default class CircularQueue<T> extends ObservableVersioned {
  frontPointer: number;
  rearPointer: number;
  items: T[];
  size: number;
  capacity: number;

  constructor(capacity = 10) {
    super();
    this.frontPointer = 0;
    this.rearPointer = 0;
    this.items = new Array(capacity);
    this.size = 0;
    this.capacity = capacity;
  }

  getSize(): number {
    return this.size;
  }

  isFull() {
    return this.size === this.capacity;
  }

  isEmpty() {
    return this.size === 0;
  }

  enqueue(item: T) {
    if (this.isFull()) {
      throw new Error("Queue Full");
    }

    this.items[this.rearPointer] = item;
    this.rearPointer += 1;
    this.rearPointer %= this.items.length;
    this.size += 1;
    this.tickVersion();
  }

  dequeue(): T {
    if (this.isEmpty()) {
      throw new Error("Queue Empty");
    }

    const item = this.items[this.frontPointer];
    this.frontPointer += 1;
    this.frontPointer %= this.items.length;
    this.size -= 1;
    this.tickVersion();
    return item;
  }
}
